omegaconf/basecontainer.py

```diff
@@ -391,16 +397,18 @@ def _set_item_impl(self, key: Any, value: Any) -> None:
 
         elif isinstance(self.__dict__["_content"], list):
             target_node = isinstance(target_node_ref, ValueNode)
+        # We use set_value if:
+        # 1. Target node is a container and the value is MISSING or None
+        # 2. Target node is a container and has an explicit ref_type
+        # 3. Target node is a valuenode. If it's an AnyNode then if the
+        #    value is a primitive type it should set.
         should_set_value = target_node_ref is not None and (
             (
-                # Target node is a container and has an special value or has an
-                # ref_type assigned
                 isinstance(target_node_ref, Container)
                 and (special_value or target_node_ref._has_ref_type())
             )
-            # if target_node is a valuenode and is not an anynode then
-            # it should set value
             or (target_node and not isinstance(target_node_ref, AnyNode))
+            or (isinstance(target_node_ref, AnyNode) and is_primitive_type(value))
```

Comment:
This got me worried.
You completely changed the logic without changing any test.
This means you are just not covering this condition in a test, and I have no way to understand what you are trying to do here and if I agree with it.

Add tests that are covering this branch.

Reply:
so, the new tests you added, are they failing with the previous version of this code?

Reply:
still waiting for an answer on for this one.

Reply:
> so, the new tests you added, are they failing with the previous version of this code?

Sorry I didn't see this comment.

They do fail under previous changes because it wraps always instead of setting value when it should.